amr-wind/equation_systems/tke/source_terms/KransAxell.cpp

```diff
@@ -0,0 +1,128 @@
+                std::max(heat_flux, 0.0) * 0.41 * x3 / std::pow(0.556, 3),
+                (2.0 / 3.0));
+            bcforcing =
+                (ustar * ustar / (0.556 * 0.556) + rans_b - tke_arr(i, j, k)) /
+                dt;
+        }
+        const amrex::Real z = problo[2] + (k + 0.5) * dx[2];
+        const amrex::Real zi =
+            std::max((z - sponge_start) / (probhi[2] - sponge_start), 0.0);
+        const amrex::Real sponge_forcing =
+            zi * zi * (tke_arr(i, j, k) - ref_tke);
+        dissip_arr(i, j, k) = std::pow(Cmu, 3) *
+                              std::pow(tke_arr(i, j, k), 1.5) /
+                              (tlscale_arr(i, j, k) + tiny);
+        src_term(i, j, k) += shear_prod_arr(i, j, k) + buoy_prod_arr(i, j, k) -
+                             dissip_arr(i, j, k) - sponge_forcing + bcforcing;
+    });
+    //! Add terrain components
+    const bool has_terrain =
+        this->m_sim.repo().int_field_exists("terrain_blank");
+    if (has_terrain) {
+        const auto* const m_terrain_blank =
+            &this->m_sim.repo().get_int_field("terrain_blank");
+        const auto* const m_terrain_drag =
+            &this->m_sim.repo().get_int_field("terrain_drag");
+        const auto& blank_arr = (*m_terrain_blank)(lev).const_array(mfi);
+        const auto& drag_arr = (*m_terrain_drag)(lev).const_array(mfi);
+        amrex::ParallelFor(
+            bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                amrex::Real terrainforcing = 0;
+                amrex::Real dragforcing = 0;
+                const amrex::Real ux = vel(i, j, k, 0);
+                const amrex::Real uy = vel(i, j, k, 1);
+                amrex::Real ustar = 0.4;
+                const amrex::Real x3 = 0.5 * dx[2];
+                if (drag_arr(i, j, k) == 1) {
+                    amrex::Real m = std::sqrt(ux * ux + uy * uy);
+                    const amrex::Real kappa = 0.41;
+                    const amrex::Real z0 = 0.1;
+                    ustar = m * kappa / std::log(x3 / z0);
+                    const amrex::Real rans_b = std::pow(
+                        std::max(heat_flux, 0.0) * 0.41 * x3 /
+                            std::pow(0.556, 3),
+                        (2.0 / 3.0));
+                    terrainforcing = (ustar * ustar / (0.556 * 0.556) + rans_b -
+                                      tke_arr(i, j, k)) /
+                                     dt;
+                }
+                if (blank_arr(i, j, k) == 1) {
+                    const amrex::Real uz = vel(i, j, k, 2);
+                    const amrex::Real m =
+                        std::sqrt(ux * ux + uy * uy + uz * uz);
+                    const amrex::Real Cd =
+                        std::min(10 / (dx[2] * m + tiny), 100 / dx[2]);
+                    dragforcing = -Cd * m * tke_arr(i, j, k, 0);
+                }
+                src_term(i, j, k) += terrainforcing + dragforcing;
```

Comment:
Let's see if we can remove some branching (not good on GPU): always combute dragforcing and terrainforcing, then do `src_term(i, j, k) += drag_arr(i, j, k) * terrainforcing + blank_arr(i, j, k)* dragforcing;` or break up the loops.